Allow user to specify a PATH to search for gem overrides

kbrock opened the pull request:
This uses a PATH to locate an override gem.

The idea behind this is a developer has a common directory where they `git clone` gems and tweak them.


I think the 2 outstanding issues are:
- What do we call the override path?
- ~~When do we not want to override a path? (maybe the developer has a gem checked out but they are overriding it with a github repo?)~~ SOLUTION: only override if no options specified (or they provide a partial path)

kbrock commented:
update:
- `s/(File.)exists/exist/g`

(annoyingly, I had added File.exists? and File.exist?)

agrare commented on lib/bundler/inject/dsl_patch.rb:
```diff
@@ -97,6 +99,10 @@ def load_bundler_d(dir)
       end
+
+      def search_path(calling_file)
+        [File.dirname(calling_file)] + ENV.fetch("BUNDLE_BUNDLER_INJECT__GEM_PATH", "").split(':')
```
We probably should specify in the README what the search order is, I think this is perfect though prioritize the local dir then in order in the BUNDLE_BUNDLER_INJECT__GEM_PATH

Fryguy replied:
Minor, but let's move any env var handling to the same place that skip_warnings? is, and wrap it in a proper method.

Fryguy replied:
Also, once you move over there, you'll see the preferable way to access the settings is with `Bundler.settings["bundler_inject.gem_path"]`.  This allows all of ENV var, .bundle/config file, global bundler config, and local bundler config.

Fryguy commented:
Please also include some README docs next to the current docs for skip warnings.

Separately, I'm not sure about `gem_path` as the setting name, but I can't think of anything better offhand...I'll think about  it.

Fryguy commented:
I tend to like explicitness with these kinds of options.  My first thought was `override_gem_default_path` (though if this is also used by ensure_gem, then that doesn't work)

Fryguy commented on spec/bundler_inject_spec.rb:
```diff
@@ -143,6 +143,30 @@
+            write_bundler_d_file <<~F
+              override_gem "ansi", :path => "the_gem"
+            F
+            bundle(:update, env: { "BUNDLE_BUNDLER_INJECT__GEM_PATH" => "#{path.dirname.to_s}/"})
```
I think the .to_s here is redundant

kbrock replied:
I removed the interpolation, but ended up needing the `to_s` again. (blew up without it)

Fryguy replied:
Right, if interpolating you don't need the to_s, but if you removed the interpolation, then you do need it.

Fryguy commented on lib/bundler/inject/dsl_patch.rb:
```diff
@@ -59,9 +59,11 @@ def remove_dependencies_and_sources(dependency)
-        args.last[:path] = File.expand_path(args.last[:path], File.dirname(calling_file))
+      def expand_gem_path(name, args, calling_file)
+        args << {} unless args.last.kind_of?(Hash)
+        path_specified = args.last[:path] || name
```
I guess _not_ setting the setting could be the answer, and if you have the setting, then we assume all overrides are path based? If so, we need to detail that nuance in the readme.

kbrock replied:
I avoided looking to into the `args`.
We now skip if version or anything is specified

kbrock commented:
update:
- only add a path if there is no version or any other argument
- use `bundler_inject.gem_path` or `$BUNDLE_BUNDLER_INJECT__GEM_PATH`
- use `File::PATH_SEPARATOR`
- using hashrocket for env
- no longer fix coverage_helper fix (another PR)

update:
- added README documentation

update:
- add Bundler::Inject.gem_path

Fryguy commented:
Can you update the PR title to reflect what you are doing here?

Fryguy commented on lib/bundler/inject.rb:
```diff
@@ -33,6 +33,24 @@ def skip_warnings?
+      #   BUNDLE_BUNDLER_INJECT__GEM_PATH=~/src:~/gems_src
+      #
+      def gem_path
+        @gem_path ||= (Bundler.settings["bundler_inject.gem_path"] || "").split(File::PATH_SEPARATOR)
```
Should we also `.map { |p| File.expand_path(p) }`, on the end of this to expand any `~`?

Fryguy commented on spec/support/helpers.rb:
```diff
@@ -92,14 +92,14 @@ def rm_global_bundler_d_dir
     end
 
-    def with_path_based_gem(source_repo)
+    def with_path_based_gem(source_repo, dir_name = "the_gem")
```
I noticed a bunch of the callers were changed (e.g. spec/bundler_inject_spec.rb:121) to set the dir_name at the caller side.  If they are all hardcoded at the caller, do we still need the default value here anymore (hard to tell if they were all changed)?

kbrock replied:
oops. I stated this below.
I think this is only called from a few spots. I will remove the default.

Fryguy commented on spec/bundler_inject_spec.rb:
```diff
@@ -117,8 +117,8 @@
-        it "with a path" do
-          with_path_based_gem("https://github.com/rubyworks/ansi") do |path|
+        it "with an absolute path" do
+          with_path_based_gem("https://github.com/rubyworks/ansi", "the_gem") do |path|
```
More out of curiosity, why was this hardcoded to the parameter default?

kbrock replied:
I had it as the default in the helper method because I wanted the interface to be backwards compatible.
In this spec, it referenced "the_gem", so I felt it was good to have this string in the spec.

We could probably drop the default.

Fryguy commented on spec/bundler_inject_spec.rb:
```diff
@@ -143,6 +143,30 @@
+            write_bundler_d_file <<~F
+              override_gem "ansi", :path => "the_gem"
+            F
+            bundle(:update, :env => {"BUNDLE_BUNDLER_INJECT__GEM_PATH" => path.dirname.to_s})
```
Can you add a spec that shows multiple paths in the ENV var, and then something appearing in _not_ the first path?  This would exercise the File::PATH_SEPARATOR code as well as that the map picks a later value.

Fryguy commented on lib/bundler/inject/dsl_patch.rb:
```diff
@@ -97,6 +100,10 @@ def load_bundler_d(dir)
         end
       end
+
+      def search_path(gemfile_dir)
```
I'm wondering if this should just be inlined into the expand_gem_path method, since it's not being reused anywhere else.  The reason is that this method will be patched into the Bundler::Dsl module, and the generic name of the method concerns me that it might conflict in the future.

kbrock commented:
update:
- rebased master (to fix conflict)
- inlined gem_path into search_path
- renamed search_path to bundler_inject_search_path (to avoid possible collision)

update:
- added spec with multiple directories

update:
- remove default name for helper method `with_path_based_gem`
- inline search_path into expand_gempath
- extract bundler_inject_gem_path -- misunderstood inline comment above

orangewolf commented:
found this plugin, which is great for a fairly odd usecase we have, but my first thought was that we'd want a custom path. this seems like a well baked solution to me! thank you for it =-)

agrare commented:
@orangewolf glad you found it useful!

> but my first thought was that we'd want a custom path

You could pass paths as an option before so for example: `override_gem "manageiq-api", :path => "~/src/manageiq/manageiq-api"` and this path could be anywhere.  This PR helps if you have a lot of overrides and they all tend to live in the same dir, you don't have to keep repeating yourself :)
